File: backend/api/sessions.py
"""
Session endpoints for viewing and managing active user sessions.

This module provides endpoints for retrieving active sessions, revoking specific
sessions, and revoking all sessions for security purposes.
"""
import logging
from datetime import datetime
from typing import List, Optional
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field
from sqlalchemy import select, update
from sqlalchemy.ext.asyncio import AsyncSession

from database import get_db
from models.session import Session

logger = logging.getLogger(__name__)

router = APIRouter()


class SessionItem(BaseModel):
    """Single session item."""

    id: str = Field(..., description="Session ID")
    user_id: str = Field(..., description="User who owns this session")
    device_name: Optional[str] = Field(None, description="User-friendly device name")
    device_type: Optional[str] = Field(None, description="Device type: desktop, mobile, tablet, or unknown")
    user_agent: Optional[str] = Field(None, description="Full user agent string")
    ip_address: Optional[str] = Field(None, description="IP address where session was created")
    location: Optional[str] = Field(None, description="Location derived from IP")
    is_active: bool = Field(..., description="Whether the session is currently active")
    expires_at: Optional[str] = Field(None, description="Session expiration timestamp")
    last_activity_at: str = Field(..., description="Timestamp of last user activity")
    created_at: str = Field(..., description="When the session was created")
    is_current: bool = Field(False, description="Whether this is the current session")


class SessionsResponse(BaseModel):
    """Response model for sessions list."""

    sessions: List[SessionItem] = Field(..., description="List of active sessions")
    total_count: int = Field(..., description="Total number of sessions")


@router.get(
    "/",
    response_model=SessionsResponse,
    tags=["Sessions"],
)
async def get_sessions(
    user_id: Optional[str] = Query(None, description="Filter by user ID"),
    device_type: Optional[str] = Query(None, description="Filter by device type"),
    is_active: Optional[bool] = Query(None, description="Filter by active status"),
    limit: int = Query(100, ge=1, le=1000, description="Maximum number of sessions to return"),
    offset: int = Query(0, ge=0, description="Number of sessions to skip for pagination"),
    db: AsyncSession = Depends(get_db),
) -> JSONResponse:
    """
    Get active sessions with filtering options.

    This endpoint retrieves active user sessions across the system, including
    device information, IP addresses, and activity timestamps. Sessions are
    returned in reverse chronological order (most recently active first).

    **Note:** This endpoint requires authentication. In production, you should
    validate the JWT token and extract the user_id to only return sessions
    belonging to the authenticated user.

    Args:
        user_id: Optional filter for specific user (requires admin privileges in production)
        device_type: Optional filter for device type (desktop, mobile, tablet, unknown)
        is_active: Optional filter for active status (true for active, false for revoked)
        limit: Maximum number of sessions to return (default: 100, max: 1000)
        offset: Number of sessions to skip for pagination (default: 0)
        db: Database session

    Returns:
        JSON response with list of sessions and total count

    Raises:
        HTTPException(401): If not authenticated
        HTTPException(403): If trying to access other user's sessions without admin privileges
        HTTPException(500): If data retrieval fails

    Examples:
        >>> import requests
        >>> response = requests.get(
        ...     "http://localhost:8000/api/sessions/?limit=10",
        ...     headers={"Authorization": "Bearer <token>"}
        ... )
        >>> response.json()
        {
            "sessions": [
                {
                    "id": "session-1",
                    "user_id": "user-1",
                    "device_name": "Chrome on Windows",
                    "device_type": "desktop",
                    "user_agent": "Mozilla/5.0...",
                    "ip_address": "192.168.1.1",
                    "location": "San Francisco, CA",
                    "is_active": true,
                    "expires_at": "2026-02-01T10:30:00Z",
                    "last_activity_at": "2026-01-31T15:45:00Z",
                    "created_at": "2026-01-31T10:30:00Z",
                    "is_current": true
                }
            ],
            "total_count": 1
        }
    """
    try:
        # TODO: In production, extract user_id from JWT token
        # TODO: Add authorization check - users can only see their own sessions unless admin
        # TODO: Get current session token to mark is_current flag

        # For now, return 401 to indicate authentication is required
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Authentication required. Please provide a valid JWT token.",
        )

        logger.info(
            f"Fetching sessions - user_id: {user_id}, device_type: {device_type}, "
            f"is_active: {is_active}"
        )

        # Build base query
        query = select(Session)

        # Apply filters
        if user_id:
            try:
                user_uuid = UUID(user_id)
                query = query.where(Session.user_id == user_uuid)
            except ValueError:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail=f"Invalid user_id format: {user_id}",
                )

        if device_type:
            valid_types = ["desktop", "mobile", "tablet", "unknown"]
            if device_type not in valid_types:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail=f"Invalid device_type: {device_type}. "
                           f"Valid types are: {', '.join(valid_types)}",
                )
            query = query.where(Session.device_type == device_type)

        if is_active is not None:
            query = query.where(Session.is_active == is_active)

        # Order by last_activity_at descending (most recent first) and apply pagination
        query = query.order_by(Session.last_activity_at.desc()).limit(limit).offset(offset)

        # Execute query
        result = await db.execute(query)
        sessions = result.scalars().all()

        # Build response data
        sessions_data = []
        for session in sessions:
            sessions_data.append({
                "id": str(session.id),
                "user_id": str(session.user_id),
                "device_name": session.device_name,
                "device_type": session.device_type,
                "user_agent": session.user_agent,
                "ip_address": session.ip_address,
                "location": session.location,
                "is_active": session.is_active,
                "expires_at": session.expires_at.isoformat() if session.expires_at else None,
                "last_activity_at": session.last_activity_at.isoformat(),
                "created_at": session.created_at.isoformat(),
                "is_current": False,  # TODO: Compare with current session token
            })

        response_data = {
            "sessions": sessions_data,
            "total_count": len(sessions_data),
        }

        logger.info(f"Retrieved {len(sessions_data)} sessions")

        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content=response_data,
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error retrieving sessions: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to retrieve sessions: {str(e)}",
        ) from e


class RevokeSessionResponse(BaseModel):
    """Response model for session revocation."""

    message: str = Field(..., description="Success message")
    session_id: str = Field(..., description="ID of the revoked session")


@router.delete(
    "/{session_id}",
    response_model=RevokeSessionResponse,
    tags=["Sessions"],
)
async def revoke_session(
    session_id: str,
    reason: Optional[str] = Query(None, description="Reason for revocation (e.g., 'user_logout', 'security_reset')"),
    db: AsyncSession = Depends(get_db),
) -> JSONResponse:
    """
    Revoke a specific session.

    This endpoint revokes a specific session by ID, effectively logging out
    the user from that device. The session is marked as inactive and a
    revocation timestamp is recorded.

    **Note:** This endpoint requires authentication. Users can only revoke
    their own sessions unless they have admin privileges.

    Args:
        session_id: ID of the session to revoke
        reason: Optional reason for revocation
        db: Database session

    Returns:
        JSON response with success message

    Raises:
        HTTPException(401): If not authenticated
        HTTPException(403): If trying to revoke another user's session without admin privileges
        HTTPException(404): If session not found
        HTTPException(400): If session is already revoked
        HTTPException(500): If revocation fails

    Examples:
        >>> import requests
        >>> response = requests.delete(
        ...     "http://localhost:8000/api/sessions/abc-123?reason=user_logout",
        ...     headers={"Authorization": "Bearer <token>"}
        ... )
        >>> response.json()
        {
            "message": "Session revoked successfully",
            "session_id": "abc-123"
        }
    """
    try:
        # TODO: In production, extract user_id from JWT token
        # TODO: Add authorization check - users can only revoke their own sessions unless admin

        # For now, return 401 to indicate authentication is required
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Authentication required. Please provide a valid JWT token.",
        )

        logger.info(f"Revoking session: {session_id}, reason: {reason}")

        # Validate session_id format
        try:
            session_uuid = UUID(session_id)
        except ValueError:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Invalid session_id format: {session_id}",
            )

        # Get the session
        query = select(Session).where(Session.id == session_uuid)
        result = await db.execute(query)
        session = result.scalar_one_or_none()

        if not session:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Session not found: {session_id}",
            )

        # Check if session is already revoked
        if not session.is_active:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Session is already revoked: {session_id}",
            )

        # TODO: Check if session belongs to current user or user is admin

        # Revoke the session
        revoke_reason = reason or "user_logout"
        await db.execute(
            update(Session)
            .where(Session.id == session_uuid)
            .values(
                is_active=False,
                revoked_at=datetime.now(),
                revoke_reason=revoke_reason,
            )
        )
        await db.commit()

        logger.info(f"Successfully revoked session: {session_id}")

        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content={
                "message": "Session revoked successfully",
                "session_id": session_id,
            },
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error revoking session: {e}", exc_info=True)
        await db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to revoke session: {str(e)}",
        ) from e


class RevokeAllResponse(BaseModel):
    """Response model for revoking all sessions."""

    message: str = Field(..., description="Success message")
    revoked_count: int = Field(..., description="Number of sessions revoked")


@router.delete(
    "/revoke-all",
    response_model=RevokeAllResponse,
    tags=["Sessions"],
)
async def revoke_all_sessions(
    user_id: str = Query(..., description="User ID to revoke all sessions for"),
    exclude_current: bool = Query(True, description="Whether to exclude the current session from revocation"),
    reason: Optional[str] = Query("security_reset", description="Reason for revocation"),
    db: AsyncSession = Depends(get_db),
) -> JSONResponse:
    """
    Revoke all sessions for a user.

    This endpoint revokes all active sessions for a specific user. This is useful
    for security purposes, such as when a user changes their password or suspects
    unauthorized access.

    By default, the current session (the one making the request) is excluded from
    revocation to avoid logging out the user who initiated the action.

    **Note:** This endpoint requires authentication. Users can only revoke their
    own sessions unless they have admin privileges.

    Args:
        user_id: User ID to revoke all sessions for
        exclude_current: Whether to exclude the current session (default: true)
        reason: Reason for revocation (default: "security_reset")
        db: Database session

    Returns:
        JSON response with success message and count of revoked sessions

    Raises:
        HTTPException(401): If not authenticated
        HTTPException(403): If trying to revoke another user's sessions without admin privileges
        HTTPException(400): If user_id format is invalid
        HTTPException(500): If revocation fails

    Examples:
        >>> import requests
        >>> response = requests.delete(
        ...     "http://localhost:8000/api/sessions/revoke-all?user_id=abc-123&reason=password_change",
        ...     headers={"Authorization": "Bearer <token>"}
        ... )
        >>> response.json()
        {
            "message": "All sessions revoked successfully",
            "revoked_count": 3
        }
    """
    try:
        # TODO: In production, extract user_id from JWT token
        # TODO: Add authorization check - users can only revoke their own sessions unless admin
        # TODO: Get current session token to exclude from revocation if exclude_current is True

        # For now, return 401 to indicate authentication is required
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Authentication required. Please provide a valid JWT token.",
        )

        logger.info(
            f"Revoking all sessions for user: {user_id}, "
            f"exclude_current: {exclude_current}, reason: {reason}"
        )

        # Validate user_id format
        try:
            user_uuid = UUID(user_id)
        except ValueError:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Invalid user_id format: {user_id}",
            )

        # TODO: Check if user_id matches current user or user is admin

        # Build query to revoke all active sessions for user
        query = (
            update(Session)
            .where(Session.user_id == user_uuid)
            .where(Session.is_active == True)
        )

        # TODO: If exclude_current is True, exclude current session token
        # if exclude_current:
        #     current_token = get_current_session_token()
        #     query = query.where(Session.token != current_token)

        # Get count before revocation
        count_query = select(Session).where(
            Session.user_id == user_uuid,
            Session.is_active == True,
        )
        # TODO: Apply exclude_current filter here too
        count_result = await db.execute(count_query)
        sessions_to_revoke = count_result.scalars().all()
        revoked_count = len(sessions_to_revoke)

        # Perform revocation
        revoke_reason = reason or "security_reset"
        query = query.values(
            is_active=False,
            revoked_at=datetime.now(),
            revoke_reason=revoke_reason,
        )

        await db.execute(query)
        await db.commit()

        logger.info(f"Successfully revoked {revoked_count} sessions for user: {user_id}")

        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content={
                "message": "All sessions revoked successfully",
                "revoked_count": revoked_count,
            },
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error revoking all sessions: {e}", exc_info=True)
        await db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to revoke all sessions: {str(e)}",
        ) from e

File: frontend/src/api/sessions.ts
/**
 * Sessions API Client
 *
 * This module provides a typed client for session management operations including
 * listing active sessions, revoking specific sessions, and revoking all sessions.
 *
 * @example
 * ```ts
 * import { sessionsClient } from '@/api/sessions';
 *
 * // List active sessions
 * const sessions = await sessionsClient.listSessions({
 *   user_id: 'user-123',
 *   device_type: 'desktop',
 *   is_active: true,
 * });
 *
 * // Revoke a specific session
 * await sessionsClient.revokeSession('session-123', 'user_logout');
 *
 * // Revoke all sessions for a user
 * const result = await sessionsClient.revokeAllSessions({
 *   user_id: 'user-123',
 *   exclude_current: true,
 *   reason: 'security_reset',
 * });
 * ```
 */

import axios, { AxiosInstance, AxiosError, AxiosResponse } from 'axios';
import {
  trackApiCall,
  type PerformanceStats,
  getPerformanceStats as getPerformanceStatsUtil,
} from '@/utils/performanceTracker';
import type {
  SessionsListResponse,
  RevokeSessionResponse,
  RevokeAllSessionsResponse,
  ApiError,
  ApiClientConfig,
} from '@/types/api';

/**
 * Default sessions API configuration
 */
const DEFAULT_CONFIG: ApiClientConfig = {
  baseURL: import.meta.env.VITE_API_URL ?? '',
  timeout: 30000, // 30 seconds for session operations
  headers: {
    'Content-Type': 'application/json',
  },
};

/**
 * Sessions API Client class
 *
 * Provides methods for session management with proper error handling,
 * type safety, and performance tracking.
 */
export class SessionsClient {
  private client: AxiosInstance;

  /**
   * Create a new sessions client instance
   *
   * @param config - Optional configuration overrides
   */
  constructor(config: ApiClientConfig = {}) {
    const finalConfig = { ...DEFAULT_CONFIG, ...config };

    this.client = axios.create(finalConfig);

    // Request interceptor
    this.client.interceptors.request.use(
      (config) => {
        // Add timestamp for performance tracking
        config.metadata = { startTime: Date.now() };

        // Add auth token if available
        const token = localStorage.getItem('auth_token');
        if (token) {
          config.headers.Authorization = `Bearer ${token}`;
        }

        return config;
      },
      (error) => {
        return Promise.reject(this.transformError(error));
      }
    );

    // Response interceptor
    this.client.interceptors.response.use(
      (response) => {
        const duration = Date.now() - (response.config.metadata?.startTime || 0);
        response.config.metadata = { ...response.config.metadata, duration };

        trackApiCall({
          endpoint: response.config.url || '',
          method: response.config.method?.toUpperCase() || 'GET',
          duration,
          status: response.status,
          success: true,
          timestamp: Date.now(),
        });

        return response;
      },
      (error) => {
        const duration = Date.now() - (error.config?.metadata?.startTime || 0);

        if (error.config) {
          trackApiCall({
            endpoint: error.config.url || '',
            method: error.config.method?.toUpperCase() || 'GET',
            duration,
            status: error.response?.status || 0,
            success: false,
            timestamp: Date.now(),
            error: error.message,
          });
        }

        return Promise.reject(this.transformError(error));
      }
    );
  }

  /**
   * Transform Axios error to standardized API error
   *
   * @param error - Axios error
   * @returns Transformed API error
   */
  private transformError(error: unknown): ApiError {
    const axiosError = error as AxiosError<{ detail?: string; message?: string }>;

    // Network error (no response)
    if (!axiosError.response) {
      if (axiosError.code === 'ECONNABORTED') {
        return {
          detail: 'Request timeout. Please check your connection and try again.',
          status: 408,
        };
      }
      return {
        detail: 'Network error. Please check your connection and try again.',
        status: 0,
      };
    }

    // Server returned error response
    const status = axiosError.response.status;
    const data = axiosError.response.data;

    // Use server's error message if available
    if (data?.detail) {
      return { detail: data.detail, status };
    }
    if (data?.message) {
      return { detail: data.message, status };
    }

    // Default error messages by status code
    const defaultMessages: Record<number, string> = {
      400: 'Invalid request. Please check your input.',
      401: 'Unauthorized. Please log in.',
      403: 'Forbidden. You do not have permission.',
      404: 'Session not found.',
      409: 'Conflict. Session may already be revoked.',
      422: 'Validation error. Please check your input.',
      500: 'Server error. Please try again later.',
    };

    return {
      detail: defaultMessages[status] || 'An unexpected error occurred.',
      status,
    };
  }

  /**
   * List active sessions with optional filters
   *
   * Retrieves active user sessions across the system, including device
   * information, IP addresses, and activity timestamps. Sessions are
   * returned in reverse chronological order (most recently active first).
   *
   * @param params - Optional query parameters for filtering
   * @returns Sessions list with total count
   * @throws ApiError if request fails
   *
   * @example
   * ```ts
   * const sessions = await sessionsClient.listSessions({
   *   user_id: 'user-123',
   *   device_type: 'desktop',
   *   is_active: true,
   *   limit: 10,
   *   offset: 0,
   * });
   * console.log(sessions.sessions.length); // Number of sessions
   * console.log(sessions.total_count); // Total count
   * ```
   */
  async listSessions(params?: {
    user_id?: string;
    device_type?: 'desktop' | 'mobile' | 'tablet' | 'unknown';
    is_active?: boolean;
    limit?: number;
    offset?: number;
  }): Promise<SessionsListResponse> {
    try {
      const queryParams: Record<string, string | number | boolean> = {
        limit: params?.limit || 100,
        offset: params?.offset || 0,
      };

      if (params?.user_id) queryParams.user_id = params.user_id;
      if (params?.device_type) queryParams.device_type = params.device_type;
      if (params?.is_active !== undefined) queryParams.is_active = params.is_active;

      const response: AxiosResponse<SessionsListResponse> = await this.client.get(
        '/api/sessions/',
        { params: queryParams }
      );
      return response.data;
    } catch (error) {
      throw this.transformError(error);
    }
  }

  /**
   * Revoke a specific session
   *
   * Revokes a specific session by ID, effectively logging out the user
   * from that device. The session is marked as inactive and a revocation
   * timestamp is recorded.
   *
   * @param sessionId - ID of the session to revoke
   * @param reason - Optional reason for revocation (e.g., 'user_logout', 'security_reset')
   * @returns Revoke session response
   * @throws ApiError if revocation fails
   *
   * @example
   * ```ts
   * await sessionsClient.revokeSession('session-123', 'user_logout');
   * // Session is now revoked
   * ```
   */
  async revokeSession(
    sessionId: string,
    reason?: string
  ): Promise<RevokeSessionResponse> {
    try {
      const queryParams: Record<string, string> = {};
      if (reason) queryParams.reason = reason;

      const response: AxiosResponse<RevokeSessionResponse> = await this.client.delete(
        `/api/sessions/${sessionId}`,
        { params: queryParams }
      );
      return response.data;
    } catch (error) {
      throw this.transformError(error);
    }
  }

  /**
   * Revoke all sessions for a user
   *
   * Revokes all active sessions for a specific user. This is useful for
   * security purposes, such as when a user changes their password or suspects
   * unauthorized access.
   *
   * By default, the current session (the one making the request) is excluded
   * from revocation to avoid logging out the user who initiated the action.
   *
   * @param params - Revocation parameters
   * @returns Revoke all sessions response with count
   * @throws ApiError if revocation fails
   *
   * @example
   * ```ts
   * const result = await sessionsClient.revokeAllSessions({
   *   user_id: 'user-123',
   *   exclude_current: true,
   *   reason: 'password_change',
   * });
   * console.log(result.revoked_count); // Number of sessions revoked
   * ```
   */
  async revokeAllSessions(params: {
    user_id: string;
    exclude_current?: boolean;
    reason?: string;
  }): Promise<RevokeAllSessionsResponse> {
    try {
      const queryParams: Record<string, string | boolean> = {
        user_id: params.user_id,
        exclude_current: params.exclude_current !== false, // Default to true
      };

      if (params.reason) queryParams.reason = params.reason;

      const response: AxiosResponse<RevokeAllSessionsResponse> = await this.client.delete(
        '/api/sessions/revoke-all',
        { params: queryParams }
      );
      return response.data;
    } catch (error) {
      throw this.transformError(error);
    }
  }

  /**
   * Get the underlying Axios instance
   *
   * This is useful for making custom requests not covered by the convenience methods.
   *
   * @returns Axios instance
   */
  getAxiosInstance(): AxiosInstance {
    return this.client;
  }

  /**
   * Get API performance statistics
   *
   * Returns performance metrics for all API calls made through this client.
   *
   * @returns Performance statistics
   */
  getPerformanceStats(): PerformanceStats {
    return getPerformanceStatsUtil();
  }
}

// Extend AxiosRequestConfig to include metadata
declare module 'axios' {
  interface AxiosRequestConfig {
    metadata?: {
      startTime?: number;
      duration?: number;
    };
  }
}

/**
 * Default sessions client instance
 *
 * Use this singleton instance for all sessions API calls.
 */
export const sessionsClient = new SessionsClient();

/**
 * Export sessions client class for custom instances
 */
export default SessionsClient;
